### PR TITLE
[Draft] Neva update to NeMo 2.0

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

Work in progress Neva's update to NeMo 2.0. In this PR we will support NeMo pretrain and fine-tuning as NeMo 1.0.

**Collection**: [multimodal, vlm]

# Changelog 
- Add mock data modules
- Re-write lazy dataset modules for conversational llava like dataset
- Rebase the implementation of Model Module of Neva to be based on MCore llava. So we can reuse many parts of the codes and all optimizations from MCore side.
- Add Neva Model Configs for NeMo 2.0 design
- Add Importer code for loading from HF
- Add example script for using NeMo 2.0 to pretrain and infer Neva.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/vlm/neva/data/lazy.py [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+                    for turn in record['conversations']:
+                        matches = re.finditer('<img src="([^"]+)"', turn['value'])
+                        for match in matches:
+                            image_name = match.group(1).split("/")[-1]
+                            image_path = os.path.join(image_folder, image_name)
+                            if not os.path.isfile(image_path):
+                                logging.warning(f"Image not found: {image_path}")
+                                continue
+                            record['image'].append(image_name)  # url
+                        turn['value'] = re.sub('<img src="([^"]+)">', IMAGE_TOKEN_INDEX, turn['value'])
```
Comment (HuiyingLi) [inline nemo/collections/vlm/neva/data/lazy.py:408]: IMAGE_TOKEN_INDEX is integer type, cannot be directly subbed into string

**Review by cuichenx — COMMENTED** [review]
some comments

File: nemo/lightning/_strategy_lib.py [hunk 2]
```diff
@@ -130,9 +130,9 @@ def set_model_parallel_attributes(model, parallelism):
     # Note: Importing nemo.lightning.pytorch.strategies creates an import cycle.
     from megatron.core.transformer.transformer_config import TransformerConfig
 
-    assert (
-        type(parallelism).__name__ == 'ParallelismConfig'
-    ), f"Expected parallelism config to be of type ParallelismConfig, but got {type(parallelism)}"
+    # assert (
+    #     type(parallelism).__name__ == 'ParallelismConfig'
+    # ), f"Expected parallelism config to be of type ParallelismConfig, but got {type(parallelism)}"
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo/lightning/_strategy_lib.py:135]: I have HF Module not using Transformer Config. https://github.com/NVIDIA/NeMo/blob/2900bae8ce90ceb2b9fffd368c5a1b6d9646529b/nemo/collections/vlm/neva/model/base.py#L174 Feel weird forcing this check

File: nemo/collections/vlm/neva/data/lazy.py (hunk 1 above)
Comment (yaoyu-33) [reply nemo/collections/vlm/neva/data/lazy.py:411]: change to "<image>"

File: nemo/collections/vlm/neva/model/base.py [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+                    patch_dim=vision_config.patch_size,
+                    add_class_token=not drop_vision_class_token,
+                    class_token_len=0 if "siglip" in vision_config.model_type else 1,
+                )
+            else:
+                self._img_seq_len = 576  # TODO(yuya): Fix hardcode
+        else:
+            self._img_seq_len = 0
+
+    def _preprocess_data(
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo/collections/vlm/neva/model/base.py:326]: this class is from mcore.
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/models/multimodal/llava_model.py#L204

We will remove this method and directly use inherited one

File: nemo/lightning/_strategy_lib.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/lightning/_strategy_lib.py:135]: @akoumpa could you check this?

**Review by cuichenx — COMMENTED** [review]
Another comment is that, for the scripts in `examples/vlm`, if they're for github tests, they should be moved to `tests/collections/vlm`. If they're meant to be the API, they should be refactored to the nemo-run format and moved to `nemo/collections/vlm/recipes`

File: nemo/collections/vlm/neva/data/lazy.py [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+                images.append(image)
+
+            if images:
+                media_tensors = torch.stack(images)
+        return media_tensors
+
+    def _apply_prompt_templates(self, source, use_plain=False):
+        conv = self.conv
+
+        roles = {"human": conv.roles[0], "gpt": conv.roles[1]}
```
Comment (HuiyingLi) [inline nemo/collections/vlm/neva/data/lazy.py:323]: For pretraining data using plain template, do we still support the jsonl format we used before, where it is "User" "Assistant" instead of human/gpt?

File: examples/vlm/neva_pretrain.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    parser = argparse.ArgumentParser(description="NEVA Model Training Script")
+
+    # Argument parsing
+    parser.add_argument("--data_path", type=str, required=True, help="Path to the dataset JSON file")
+    parser.add_argument("--image_folder", type=str, required=True, help="Path to the image folder")
+    parser.add_argument("--log_dir", type=str, required=True, help="Directory for logging and checkpoints")
+    parser.add_argument("--language_model_path", type=str, required=True, help="Path to the pretrained language model")
+    parser.add_argument("--devices", type=int, required=False, default=1)
+    parser.add_argument("--tp_size", type=int, required=False, default=1)
+    parser.add_argument("--projector_type", type=str, required=False, default="mlp")
```
Comment (HuiyingLi) [inline examples/vlm/neva_pretrain.py:156]: mlp2x_gelu, regex doesn't capture mlp

File: nemo/collections/vlm/neva/data/lazy.py (hunk 4 above)
Comment (yaoyu-33) [reply nemo/collections/vlm/neva/data/lazy.py:323]: I think I manually converted the user/assitant to human/gpt before for jsonl files

**Review by hemildesai — APPROVED** [review]
